### ISSUE TITLE
Bug in FullyConnected Shape Inference

### ISSUE THREAD

**Issue by pluskid** [dmlc/mxnet#1778]
The `FullyConnected` layer use mshadow's [`FlatTo2D`](https://github.com/dmlc/mshadow/blob/master/mshadow/tensor_blob.h#L179) when doing shape inference (see [here](https://github.com/dmlc/mxnet/blob/master/src/operator/fully_connected-inl.h#L162)). However, the behavior of `FlatTo2D` does not match the needs here. Essentially, `FlatTo2D` flat `(2,3,4,5)` to `(2*3*4, 5)`, while we need `(2, 3*4*5)` in the shape inference of fully connected layers. It seems `FlatTo2D` is used in many places, so I would like to post this issue here just in case there might have some other potential issues.

The current bug is that unless a `Flatten` layer is explicitly inserted between a convolutional 4D tensor and a fully connected layer, the shape inference will be wrong. Here is an example:

```python
import mxnet as mx

def get_lenet():
    """
    LeCun, Yann, Leon Bottou, Yoshua Bengio, and Patrick
    Haffner. "Gradient-based learning applied to document recognition."
    Proceedings of the IEEE (1998)
    """
    data = mx.symbol.Variable('data')
    # first conv
    conv1 = mx.symbol.Convolution(data=data, kernel=(5,5), num_filter=20)
    tanh1 = mx.symbol.Activation(data=conv1, act_type="tanh")
    pool1 = mx.symbol.Pooling(data=tanh1, pool_type="max",
                              kernel=(2,2), stride=(2,2))
    # second conv
    conv2 = mx.symbol.Convolution(data=pool1, kernel=(5,5), num_filter=50)
    tanh2 = mx.symbol.Activation(data=conv2, act_type="tanh")
    pool2 = mx.symbol.Pooling(data=tanh2, pool_type="max",
                              kernel=(2,2), stride=(2,2))
    # first fullc
    #flatten = mx.symbol.Flatten(data=pool2)
    flatten = pool2
    fc1 = mx.symbol.FullyConnected(data=flatten, num_hidden=500)
    tanh3 = mx.symbol.Activation(data=fc1, act_type="tanh")
    # second fullc
    fc2 = mx.symbol.FullyConnected(data=tanh3, num_hidden=10)
    # loss
    lenet = mx.symbol.SoftmaxOutput(data=fc2, name='softmax')
    return lenet

lenet = get_lenet()
arg_names = lenet.list_arguments()
arg_s, aux_s, out_s = lenet.infer_shape(data=(100,1,100,100), softmax_label=(100,))

for n, s in zip(arg_names, arg_s):
    print n, s
```

which gives
```python
data (100L, 1L, 100L, 100L)
convolution2_weight (20L, 1L, 5L, 5L)
convolution2_bias (20L,)
convolution3_weight (50L, 20L, 5L, 5L)
convolution3_bias (50L,)
fullyconnected2_weight (500L, 22L)  # NOTE this 22 is wrong
fullyconnected2_bias (500L,)
fullyconnected3_weight (10L, 500L)
fullyconnected3_bias (10L,)
softmax_label (100L,)
```

Related issue dmlc/MXNet.jl#69

**Comment by tqchen** [comment]
@antinucleon can you look into this?

**Comment by tqchen** [comment]
I agree this could be a bug we overlooked, we should use  https://github.com/dmlc/mshadow/blob/master/mshadow/tensor.h#L133 instead

**Comment by pluskid** [comment]
I think it might worth checking other use of `FlatTo2D` to make sure all are anticipating the expected behavior.

**Comment by tqchen** [comment]
I see. Most operator as far as I know uses FlatTo2D for elementwise ops, so they should be fine

**Comment by tqchen** [comment]
@pluskid can you submit a PR

**Comment by tqchen** [comment]
This is fixed by #1824

**Comment by pluskid** [comment]
Thanks a lot!
